File: Binary Search/searchInRotatedSortedArray.cpp
/*
	Given the array nums after the possible rotation and an integer target, return the index of target if it is in nums, or -1 if it is not in nums.
	You must write an algorithm with O(log n) runtime complexity.

	Input: nums = [4,5,6,7,0,1,2], target = 0
	Output: 4
	
	Input: nums = [4,5,6,7,0,1,2], target = 3
	Output: -1
	
	Input: nums = [1], target = 0
	Output: -1

	Constraints:
	>  1 <= nums.length <= 5000
	>  -104 <= nums[i] <= 104
	>  All values of nums are unique.
	>  nums is an ascending array that is possibly rotated.
	>  -104 <= target <= 104

	APPROACH
	1. Find the index at which the array has been rotated with the help of Binary Search.
	2. Store the index of rotation as INDX and find the range where target might be found using the following comparision:
			RANGE = {INDX+1, HIGH} if TARGET < NUMS[LOW]
			RANGE = {LOW, INDX-1}  if TARGET > NUMS{HIGH]
	3. Perform Binary Search for TARGET in the required range.
	4. If target is not found return -1.

	TIME COMPLEXITY : O(NlogN)  SPACE COMPLEXITY: O(1)
*/
class Solution {
public:
    int search(vector<int>& nums, int target) {
        // Initialize Variable for later Usage
        int n=nums.size();
        int low=0;
        int high=n-1;

        //Find the Rotation Point in the Rotated Array
        while(low<=high){
            int mid=(low+high)/2;
            if(mid==0 || mid==n-1){
                low=mid;
                break;
            }
            if(nums[mid-1]>nums[mid] && nums[mid+1]>nums[mid]){
                low=mid;
                break;
            }
            else if(nums[mid]>nums[low] && nums[mid]>nums[high]){
                low=mid+1;
            }
            else{
                high=mid-1;
            }
        }

        // Re-initialize Variables Needed
        int indx=low;
        low=0, high=n-1;
        if(target==nums[indx]){
            return indx;
        }
        else if(target>nums[high]){
            high=indx-1;
        }
        else if(target<nums[low]){
            low=indx+1;
        }

        // Binary  Search for Target in range low-high
        while(low<=high){
            int mid=(low+high)/2;
            if(nums[mid]==target){
                return mid;
            }
            else if(target>nums[mid]){
                low=mid+1;
            }
            else{
                high=mid-1;
            }
        }

        // If target not found return -1
        return -1;
    }
};

File: Binary Search/Search_in_Rotated_Sorted_Array.java


/* Problem:
  There is an integer array nums sorted in ascending order (with distinct values).
  Prior to being passed to your function, nums is possibly rotated at an unknown pivot index k (1 <= k < nums.length) such that the resulting array is [nums[k], nums[k+1], ..., nums[n-1], nums[0], nums[1], ..., nums[k-1]] (0-indexed). For example, [0,1,2,4,5,6,7] might be rotated at pivot index 3 and become [4,5,6,7,0,1,2].
  Given the array nums after the possible rotation and an integer target, return the index of target if it is in nums, or -1 if it is not in nums.
  You must write an algorithm with O(log n) runtime complexity.
  
Sample Input-Output:

  Example 1:
  Input: nums = [4,5,6,7,0,1,2], target = 0
  Output: 4
  
  Example 2:
  Input: nums = [4,5,6,7,0,1,2], target = 3
  Output: -1
  
  Example 3:
  Input: nums = [1], target = 0
  Output: -1
  
Constraints:
  1 <= nums.length <= 5000
  -104 <= nums[i] <= 104
  All values of nums are unique.
  nums is an ascending array that is possibly rotated.
  -104 <= target <= 104
  
Approach: 
  This approach has a time complexity of O(log n).
  
*/
  
class Search_in_Rotated_Sorted_Array {
    public int search(int[] nums, int target) {
        int start = 0, end = nums.length - 1;
        int mid = (start + end) / 2;    //finding the mid element
        while (start <= end) {    //until the target element is found or the start pointer becomes greater than the end pointer
            mid = (start + end) / 2;
            if (target == nums[mid]) {    //comparing it with the target element.
                return mid;   //If they are equal, it returns the mid index       
            }
            if (nums[start] <= nums[mid]) {   //If the left half of the array is sorted
                if (nums[start] <= target && nums[mid] >= target) {   // checking if the target lies between the start and the mid
                    end = mid - 1;    // updating the end pointer accordingly.
                } else {    //checking if the target lies between mid and end
                    start = mid + 1;    //updating the start pointer accordingly.
                }
            } else {     // If the right half of the array is sorted
                if (nums[end] >= target && nums[mid] <= target) {   //checking if the target lies between mid and end
                    start = mid + 1;    //updating the start pointer accordingly.
                } else {    //checking if the target lies between start and mid
                    end = mid - 1;    //updating the end pointer accordingly.
                }
            }
        }
        return -1;    //returns -1 if element not found
    }
}
